### PR TITLE
Create private deployment

### PR DESCRIPTION
**Pull request by jasonmule**
#859

- [ ] Add API tests

[Review on Reviewable](https://reviewable.io/reviews/ushahidi/platform/866)

### CODE REVIEW

File: application/classes/Ushahidi/Core.php [hunk 1]
```diff
@@ -51,11 +51,14 @@ public static function init()
 		});
 
 		// Ratelimiter config settings
-
 		$di->set('ratelimiter.config', function() use ($di) {
 			return Kohana::$config->load('ratelimiter')->as_array();
 		});
 
+		// Private deployment config settings
+		$di->set('site.private', function() use ($di) {
```
Comment (rjmackay) [inline application/classes/Ushahidi/Core.php]: :+1:

File: src/Core/Tool/Authorizer/ContactAuthorizer.php [hunk 2]
```diff
@@ -31,12 +32,20 @@ class ContactAuthorizer implements Authorizer
...
+	use PrivateDeployment;
+
 	/* Authorizer */
 	public function isAllowed(Entity $entity, $privilege)
 	{
 		// These checks are run within the user context.
 		$user = $this->getUser();
 
+		// Only logged in users have access if the deployment is private
+		if ($this->isPrivate() and !$user->getId()) {
```
Comment (rjmackay) [inline src/Core/Tool/Authorizer/ContactAuthorizer.php]: I think you can do `this->getUserId` but I think we should probably add something like `->hasUser` to UserContext so we're not manually checking the id (though we already do that elsewhere)

Comment (rjmackay) [inline src/Core/Tool/Authorizer/ContactAuthorizer.php]: In fact, given you're doing this check everywhere, maybe alias it into the PrivateDeployment trait too. ie.
Add `canAccessDeployment` (or a better name) and do
```
if (! $this->canAccessDeployment()) {
  return false;
}
```
or maybe thats overkill?

File: src/Init.php [hunk 3]
```diff
@@ -301,6 +301,7 @@ function feature($name)
 $di->setter['Ushahidi\Core\Usecase\Form\VerifyStageLoaded']['setStageRepository']
 	= $di->lazyGet('repository.form_stage');
 $di->setter['Ushahidi\Core\Traits\Event']['setEmitter'] = $di->lazyNew('League\Event\Emitter');
+$di->setter['Ushahidi\Core\Traits\PrivateDeployment']['setPrivate'] = $di->lazyGet('site.private');
```
Comment (rjmackay) [inline src/Init.php]: Nice and clean :+1:

**Comment by rjmackay** [comment]
Authorizer could be a little clean but this looks good.
However we really need some tests for this.

File: src/Core/Tool/Authorizer/ContactAuthorizer.php (hunk 2 above)
Comment (jasonmule) [inline src/Core/Tool/Authorizer/ContactAuthorizer.php]: Much nicer.

**Comment by rjmackay** [comment]
Changes look good. Need some test then good to land.

**Comment by rjmackay** [comment]
Reviewed 4 of 15 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/Core/Tool/Authorizer/UserAuthorizer.php, line 85 \[r2\]](https://reviewable.io:443/reviews/ushahidi/platform/866#-K7xAr6aanRIbpyZ1m0O:-K7xAr6aanRIbpyZ1m0P:1575756833) ([raw file](https://github.com/ushahidi/platform/blob/ea7a2b4c019e3aadcc5f414de605835566f9f155/src/Core/Tool/Authorizer/UserAuthorizer.php#L85)):*
We should add tests to make sure users can't register on private deployments.
Otherwise it'd be too easy to break this in future updates to this Authorizer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ushahidi/platform/866)*

**Comment by rjmackay** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ushahidi/platform/866)*

**Comment by rjmackay** [comment]
Looks good. But make sure to merge to develop, not master

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/ushahidi/platform/866#-:-KAahrYhMea_s4OFtIAP:755787280)*

**Pull request closed without merging**